lib/resty/core/socket/tcp.lua

```diff
@@ -0,0 +1,173 @@
+-- Copyright (C) by OpenResty Inc.
+
+
+local base = require "resty.core.base"
+local ffi = require "ffi"
+
+
+local C = ffi.C
+local ffi_str = ffi.string
+local ffi_gc = ffi.gc
+local FFI_ERROR = base.FFI_ERROR
+local FFI_DONE = base.FFI_DONE
+local FFI_OK = base.FFI_OK
+local FFI_AGAIN = base.FFI_AGAIN
+local FFI_NO_REQ_CTX = base.FFI_NO_REQ_CTX
+local get_request = base.get_request
+local error = error
+local assert = assert
+local type = type
+local pcall = pcall
+local select = select
+local co_yield = coroutine._yield
+
+
+ffi.cdef[[
+typedef struct ngx_http_lua_socket_tcp_upstream_s
+    ngx_http_lua_socket_tcp_upstream_t;
+
+int ngx_http_lua_ffi_socket_tcp_sslhandshake(ngx_http_request_t *r,
+    ngx_http_lua_socket_tcp_upstream_t *u, void *sess,
+    int enable_session_reuse, ngx_str_t *server_name, int verify,
+    int ocsp_status_req, void *chain, void *pkey, char **errmsg);
+
+int ngx_http_lua_ffi_socket_tcp_get_sslhandshake_result(ngx_http_request_t *r,
+    ngx_http_lua_socket_tcp_upstream_t *u, void **sess, char **errmsg,
+    int *openssl_error_code);
+
+void ngx_http_lua_ffi_ssl_free_session(void *sess);
+]]
+
+
+local SOCKET_CTX_INDEX = 1
+
+
+local errmsg = base.get_errmsg_ptr()
+local session_ptr = ffi.new("void *[1]")
+local server_name_str = ffi.new("ngx_str_t[1]")
+local openssl_error_code = ffi.new("int[1]")
+
+local function setclientcert(self, cert, pkey)
+    if not cert and not pkey then
+        self.client_cert = nil
+        self.client_pkey = nil
+        return
+    end
+
+    if not cert or not pkey then
+        error("client certificate must be supplied with corresponding " ..
+              "private key", 2)
+    end
+
+    if type(cert) ~= "cdata" then
+        error("bad client cert type", 2)
+    end
+
+    if type(pkey) ~= "cdata" then
+        error("bad client pkey type", 2)
+    end
+
+    self.client_cert = cert
+    self.client_pkey = pkey
+end
+
+
+local function sslhandshake(self, reused_session, server_name, ssl_verify,
+    send_status_req, ...)
+
+    local n = select("#", ...)
+    if not self or n > 1 then
+        error("ngx.socket sslhandshake: expecting 1 ~ 5 arguments " ..
+              "(including the object), but seen " .. (self and 5 + n or 0))
```

Comment:
What's the reason for the `or 0` portion? Should it just be `5 + n`?

Reply:
done

Reply:
It not a bug, see t/129-ssl-socket.t  test 31 in lua-nginx-module.